경북대 FE 이창목 - week2_step2 [주문하기 페이지 구현하기]

eelkom opened the pull request:
## PR

### ✨ 작업 내용

- [x] 재사용 가능성이 높은 컴포넌트, Hooks가 보이면 재사용성을 고려하여 만들어요.
- [x] 별도의 From, Validation 라이브러리를 사용하지 않고 useState로만 구현해요.
1. 로그인 정보
- [x] 로그인 정보를 어디서든 사용할 수 있게 해요.
- [x] 새로고침을 하더라도 로그인 정보가 유지되도록 해요.
- [x] 로그인 한 유저는 로그인 버튼 클릭 시 마이페이지(/my)로 연결돼요.
- [x] 마이페이지는 로그인 한 사람만 접근 가능해요. 로그인 하지 않은 사람이 접속하면 로그인 페이지로 연결해요.
- [x] 마이페이지에서 로그아웃을 할 수 있어요. 로그아웃 시 로그인(/login) 페이지로 연결돼요.


2. 주문하기
- [x] 상품 아이템을 클릭하면 주문하기 페이지로 연결돼요.
- [x] 만약 로그인이 되어있지 않다면 로그인을 유도하고, 로그인 성공 시 주문하기 페이지로 접근 가능하게 해요.
- [x] [ 참고 URL](https://kakaotech-mission2-order-step2.pages.dev/)을 살펴보고 주문하기에 필요한 기능을 구현해요.
- [x] 버튼 클릭 시 아래 유효성 검사 조건이 충족되지 않으면 각 폼 필드에 안내 메시지를 보여줘요.

  메시지는 반드시 입력 되어야 해요.
  보내는 사람 이름이 반드시 입력 되어야 해요.
  받는 사람 이름이 반드시 입력 되어야 해요.
  받는사람 전화번호가 반드시 입력되고 전화번호 규칙에 맞아야 해요. (01012341234)
  수량은 1개 이상이어야 해요.


### 🔍 참고 사항

안녕하세요. 경북대 FE 5팀 이창목 입니다.

eelkom committed:
feat: onBlur 로직 추가

eelkom committed:
feat: context api 를 통해 로그인 정보 전역 사용

eelkom committed:
feat: 마이 페이지 구현

eelkom committed:
feat: 조건부 라우팅 구현 (로그인 상태에 따른)

eelkom committed:
refactor: 컴포넌트 내부에서 직접 navigate

eelkom committed:
feat: login 시 id, pw 컨텍스트의 userInfo에 업데이트

eelkom committed:
feat: 닉네임 표시

eelkom committed:
feat: OrderPage 구현(WIP)

- 상품 아이템을 클릭하면 주문하기 페이지로 연결돼요.
- 만약 로그인이 되어있지 않다면 로그인을 유도하고, 로그인 성공 시 주문하기 페이지로 접근 가능하게 해요.
- 참고 URL을 살펴보고 주문하기에 필요한 기능을 구현해요.

TODO:
- 버튼 클릭 시 아래 유효성 검사 조건이 충족되지 않으면 각 폼 필드에 안내 메시지를 보여줘요.
메시지는 반드시 입력 되어야 해요.
보내는 사람 이름이 반드시 입력 되어야 해요.
받는 사람 이름이 반드시 입력 되어야 해요.
받는사람 전화번호가 반드시 입력되고 전화번호 규칙에 맞아야 해요. (01012341234)
수량은 1개 이상이어야 해요.

eelkom committed:
refactor: MessageCardSection 리팩토링 (SelectedCard 영역 추가)

eelkom committed:
feat: MessageCard type 추가

eelkom committed:
refactor: useInput 공용 훅으로 분리

eelkom committed:
refactor: OrderPage 내부 컴포넌트 리팩토링

Leejin-Yang commented on src/contexts/UserInfoContext.tsx:
```diff
@@ -0,0 +1,49 @@
+  const login = (email: string, password: string) => {
+    setUserInfo({ email, password });
+    sessionStorage.setItem('email', email);
+    sessionStorage.setItem('password', password);
```
세션스토리지에 패스워드를 저장하고 있는데,
보안상 문제가 될 수 있어보여요

eelkom replied:
추후 반영하겠습니다!

Leejin-Yang commented on src/hooks/useInput.ts:
```diff
@@ -0,0 +1,40 @@
+type FieldType = 'email' | 'password' | 'text' | 'number';
+
+const useInput = (type: FieldType) => {
+  let validator: (value: string) => string;
+
+  if (type === 'email') {
+    validator = validateEmail;
+  } else if (type === 'password') {
+    validator = validatePassword;
+  } else if (type === 'text') {
+    validator = validateText;
+  } else if (type === 'number') {
+    validator = validateNumber;
+  }
```
이 부분은 훅 밖으로 뺄 수 있어보여요!
그리고 이런 분기문들은 조건이 추가될수록 길어질 수 있는데요, 객체를 활용해 리팩터링 할 수 있어보여요!

Leejin-Yang commented on src/page/OrderPage/components/OrderButton.tsx:
```diff
@@ -0,0 +1,29 @@
+const OrderButton = () => {
+  return (
+    <OrderButtonContainer>
+      29000원 주문하기
```
수량이 변경되면 값이 변경되겠네요 😊

Leejin-Yang commented on src/page/OrderPage/components/ProductInfo.tsx:
```diff
@@ -0,0 +1,81 @@
+          src="https://st.kakaocdn.net/product/gift/product/20231030175450_53e90ee9708f45ffa45b3f7b4bc01c7c.jpg"
+        />
+        <div>
+          <p>BBQ 양념치킨+크림치즈볼+콜라1.25L</p>
+          <p>BBQ</p>
+          <p>
+            <span>상품가 </span>29000원
+          </p>
```
이 정보들은 데이터에서 가져와서 작성해보면 어떨까요?

eelkom replied:
useParams 를 통해 id 값에 해당하는 데이터를 가져와 수정했습니다!

Leejin-Yang commented on src/routes/RequireAuth.tsx:
```diff
@@ -0,0 +1,17 @@
+import { ROUTES } from './routes';
+import { type PropsWithChildren } from 'react';
+
+const RequireAuth = ({ children }: PropsWithChildren) => {
```
인증이 필요한 경로는 잘 분리해주셨네요 👍

Leejin-Yang commented on src/routes/Routers.tsx:
```diff
@@ -1,14 +1,26 @@
+          </RequireAuth>
+        }
+      />
+      <Route path={ROUTES.ORDER} element={<OrderPage />} />
```
주문 페이지도 RequireAuth로 감싸줄 수 있지 않을까요?

eelkom replied:
추후 반영하겠습니다!

Leejin-Yang commented on src/page/LoginPage/index.tsx:
```diff
@@ -49,15 +51,20 @@ const Button = styled.button`
+    if (!username.isValid || !password.isValid) return;
+
+    login(username.value, password.value);
+    navigate(ROUTES.MY);
```
로그인 후에 뒤로가기를 누르면 로그인 페이지로 넘어가게 되는데, 이 부분도 수정할 수 있어보여요 😊

eelkom replied:
`replace: true` 옵션을 사용하면 히스토리 스택의 상단을 새 페이지가 차지. 
로그인 후 로그인 페이지의 기록 대신 이동한 마이페이지가 스택의 상단을 차지하고 뒤로가기를 누를 시 로그인 페이지 이전 페이지로 이동한다고 이해했습니다!

Leejin-Yang commented on src/hooks/useInput.ts:
훅을 잘 구성해주셨네요 👍
범용적으로 사용하기 위해 작성하신걸로 보이는데, 사실 조금 제한적인 역할을 하고 있을 수 있어요.
validator가 특정 필드에 종속되어 있어서 필드가 늘어날수록 훅 내부에서 모든 validator를 알아야 하는 것으로 보여요
현재 구조에서는 validator를 인자로 받아서 사용하는 방식을 고려해볼 수 있어요.

물론 지금 단계에서는 현재 방식으로도 사용할 수 있지만, 확장성을 고려해보는 것도 앞으로 범용적인 훅을 작성할 때 좋아보여요 😊

eelkom replied:
validate 파일 따로 분리 시켜 훅에서는 getValidator(type) 함수를 통해 type에 맞는 유효성 검사 함수를 불러오는 방식으로 리팩토링 했습니다!

Leejin-Yang commented on src/hooks/useInput.ts:
```diff
@@ -0,0 +1,40 @@
+    setError(validator(targetValue));
+  };
+
+  const onBlur = (e: React.ChangeEvent<HTMLInputElement>) => {
```
Blur의 이벤트는 ChangeEvent 일까요? 힌트를 드리자면 VSCode에서 onBlur 속성에 마우스를 올려보면 확인할 수 있습니다 😊

eelkom replied:
FocusEvent로 확인했습니다!

Leejin-Yang reviewed (changes requested):
창목님 안녕하세요 👋

2단계는 구현할 부분이 많아서, 많이 힘드셨을거 같은데.. 고생하셨습니다 👍
진행해주신 부분까지는 코멘트 남겨두었고, 아직 진행중인 부분까지 완료해서 다시 요청 보내주세요!
진행하다가 고민되는 부분이나 궁금한 점은 언제든 코멘트나 DM 남겨주세요 🙌

eelkom committed:
feat: OrderPage 기능 구현 및 리팩토링

- OrderPage 전체 기능 구현
- OrderPage 유효성 검사 로직 추가
- OrderPage 모든 요소 styled-component로 변경
- useInput 훅 분리
- useCheckAmount 훅 추가
- 타입 에러 수정

eelkom committed:
refactor: validate 함수 통합

eelkom committed:
chore: 파일 확장자 수정

eelkom committed:
feat: id 값에 따른 라우팅 및 data 불러오기 구현

eelkom commented:
과제 수행하면서 작성했던 학습 일지입니다. 
많은 내용은 없지만 리뷰 하실 때 도움이 될까 해서 첨부합니다!

- https://www.notion.so/teamsparta/3-22a2dc3ef51481c4b186e31176ddb56b
- https://www.notion.so/teamsparta/2-22a2dc3ef51481e9b05de555b8dbfb52
- https://www.notion.so/teamsparta/2-22a2dc3ef5148162bfadc850171ae72f

현재 useInput 훅이 OrderPage 내에 1개, LoginPage 내에 1개 존재하는데 기능이 비슷한 부분이 있어 하나로 합쳐보려고 했지만 return 값과 내부 함수의 기능이 미묘하게 다르다고 느껴서 두 개로 나누어둔 상태입니다. 
하나로 합치게 된다면 조건에 따라 onChange 내부 로직이 달라지도록 해야 할 것 같은데 이런 경우에는 어떻게 처리 하는게 좋을까요?

eelkom committed:
feat: 주문 페이지 접근 시 로그인 여부 확인 기능 리팩토링

- GiftRanking 컴포넌트에서 로그인 여부 확인 로직 제거
- Routers.tsx에서 OrderPage에 RequireAuth 적용하여 로그인된 사용자만 접근 가능하도록 변경

eelkom committed:
chore: useInput 확장자 변경

eelkom committed:
fix: RankItem(상품) 클릭 시 로그인 페이지 유도 에러 수정

- Routers 에서 RequireAuth로 분기 처리 후 navigate(ROUTES.LOGIN)
  사제하니 에러 발생
- 이유:
	RequireAuth 컴포넌트의 로직은 라우트에 접근하려고 할 때 작동
	GiftRanking.tsx의 handleItemClick 함수는 RankItem을 클릭했을 때 즉시 실행
	isLoggedIn 상태를 확인하고, 로그인되어 있지 않으면 navigate(ROUTES.LOGIN)을 직접 호출하여 로그인 페이지로 이동
	GiftRanking 컴포넌트에서 이미 로그인 페이지로 리다이렉트가 발생했기 때문에, OrderPage 라우트로의 접근 자체가 시도되지 않음
	-> OrderPage를 감싸고 있는 RequireAuth 컴포넌트의 로직은 실행될 기회가 없는 것

eelkom commented:
위 에러에 대해 제미나이에게 '두 곳에서  조건부 처리를 하는 것이 낭비 인가?' 라는 질문 하였는데 아래와 같은 답변을 보여주었습니다.
아래와 같이 이해하면 되는 부분일까요?

아닙니다, 두 곳에서 분기 처리를 하는 것이 낭비라고 단정하기는 어렵습니다. 오히려 사용자 경험(UX)과 보안/견고성
  측면에서 각각 다른 목적을 가집니다.

   1. `GiftRanking.tsx` (클릭 이벤트 핸들러): 사용자 경험 개선
       * RankItem을 클릭하는 시점에 바로 로그인 여부를 확인하고, 로그인되어 있지 않다면 즉시 로그인 페이지로
         유도합니다.
       * 이는 사용자가 보호된 페이지(OrderPage)로 이동하기 전에 필요한 조치(로그인)를 취하도록 안내하여, "클릭 ->
         페이지 이동 시도 -> 다시 로그인 페이지로 리다이렉트"와 같은 불필요한 단계를 줄여줍니다.
       * 사용자에게 "이 기능을 사용하려면 로그인이 필요합니다"라는 명확한 피드백을 즉각적으로 제공하여 더 나은
         사용자 경험을 제공합니다.


   2. `RequireAuth.tsx` (라우트 가드): 보안 및 견고성 확보
       * OrderPage 라우트 자체에 대한 접근을 보호하는 역할을 합니다.
       * 사용자가 RankItem 클릭과 같은 UI 상의 경로를 통하지 않고, 직접 URL을 입력하거나 북마크를 통해
         `OrderPage`에 접근하려고 할 때 로그인 여부를 확인하고 로그인 페이지로 리다이렉트합니다.
       * 이는 애플리케이션의 특정 페이지가 항상 인증된 사용자에게만 접근 가능하도록 보장하는 보안 계층입니다.

  결론적으로,
   * GiftRanking.tsx의 로직은 특정 UI 액션에 대한 즉각적인 사용자 안내를 위한 것이고,
   * RequireAuth.tsx는 라우트 자체에 대한 접근 제어 및 보안을 위한 것입니다.

eelkom committed:
feat: replace: true 옵션 적용

- 로그인 후 뒤로가기 클릭 시 로그인 페이지 접근되지 않도록 수정

eelkom committed:
feat: pw 보안을 위한 token 설정

eelkom commented on src/page/LoginPage/index.tsx:
```diff
@@ -58,14 +59,16 @@ const LoginPage = () => {
     if (!username.isValid || !password.isValid) return;
-
-    login(username.value, password.value);
+    const token = await fakeAuthApi(username.value, password.value);
```
로그인 버튼 클릭 시 token을 생성해 password가 그대로 입력되는 부분을 수정했습니다.
token을 어떻게 생성하는 fakeAuthApi 함수의 경우 어떻게 구현해야 할지 감이 안 잡혀서 gpt의 도움을 받았습니다.

Leejin-Yang commented on src/page/LoginPage/components/InputField.tsx:
```diff
@@ -38,6 +38,7 @@ const P = styled.p`
 export interface InputHook {
   value: string;
   onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
+  onBlur: (e: React.ChangeEvent<HTMLInputElement>) => void;
   isValid: boolean;
   error?: string;
   [key: string]: unknown;
```
다음 단계에서는 이 타입과 InputProps를 개선해보면 좋아보입니다.
필요한 부분만 props로 넘겨주고, input의 기본 속성들은 어떻게 가져올 수 있는지 고민해보면서 시도해보면 좋아보여요 😊

Leejin-Yang commented on src/page/LoginPage/index.tsx:
```diff
@@ -49,16 +52,23 @@ const Button = styled.button`
+    login(username.value, token);
+    navigate(ROUTES.MY, { replace: true });
   };
+  // const token = await fakeAuthApi(username.value, password.value);
+  
```
불필요한 주석은 삭제 부탁드려요 😊

Leejin-Yang commented on src/page/MyPage/index.tsx:
```diff
@@ -0,0 +1,35 @@
+
+  return (
+    <Container>
+      <form onSubmit={handleSubmit}>
+        <h1>마이페이지</h1>
+        <p>안녕하세요!</p>
+        <p>이메일 주소는 {userInfo.email} 입니다.</p>
+        <button>로그아웃</button>
+      </form>
```
로그아웃은 제출하는 입력 필드가 없으니, 이 부분은 클릭이벤트로 해도 괜찮아보이네요 😊

Leejin-Yang commented on src/page/LoginPage/utils/fakeAuthApi.ts:
```diff
@@ -0,0 +1,13 @@
+export const fakeAuthApi = (email: string, password: string): Promise<string> =>
+  new Promise((resolve, reject) => {
+    setTimeout(() => {
+      if (!email || !password) {
+        reject(new Error('이메일과 비밀번호를 모두 입력하세요.'));
+        return;
+      }
+
+      // “email:timestamp”를 base64 로 인코딩해 토큰처럼 사용
+      const token = btoa(`${email}:${Date.now()}`);
+      resolve(token);
+    }, 300); // 네트워크 지연 흉내
+  });
```
좋은 시도로 보입니다 👍
통신까지 한다면 이제 진정한 프론트엔드 개발을 하는거라고 생각해요. 그리고 Promise 개념도 더 학습해보시면 좋아보입니다 😊

Leejin-Yang commented on src/page/OrderPage/index.tsx:
```diff
@@ -0,0 +1,61 @@
+        amountHook={receiverAmount}
+      />
+      <ProductInfo index={index} />
+      <OrderButton onClick={handleClick} index={index} />
```
이 영역은 form으로 볼 수 있겠네요 😊

Leejin-Yang commented on src/page/OrderPage/components/MessageInput.tsx:
```diff
@@ -0,0 +1,43 @@
+      <Textarea placeholder="메시지를 입력해주세요." defaultValue="축하해요."></Textarea>
+    </MessageInputContainer>
+  );
+};
```
요구사항 중에 메시지도 제출 해야하는 걸로 알고 있는데, 이 부분이 누락된 걸로 보이네요.!

Leejin-Yang commented on src/page/OrderPage/index.tsx:
```diff
@@ -0,0 +1,61 @@
+const OrderPage = () => {
+  const { id } = useParams<{ id: string }>();
+  const index = Number(id);
+  console.log(id);
```
콘솔 로그도 제출 전에 남아있는지 확인 부탁드려요 😊

Leejin-Yang commented on src/utils/validate.ts:
```diff
@@ -0,0 +1,61 @@
+};
+
+type FieldType = 'email' | 'password' | 'text' | 'phoneNumber';
+const validators = {
+  email: validateEmail,
+  password: validatePassword,
+  text: validateText,
+  phoneNumber: validatePhoneNumber,
+  amount: validateAmount,
+} as const;
```
주문과 로그인이 하나의 유효성 검사 파일 있는데요, 기능에 맞게 파일을 분리해 관리하기 쉽게 하는건 어떻게 생각하시나요?

Leejin-Yang reviewed (changes requested):
피드백에 대해서 그리고 개인적으로도 많이 고민을 하신 것이 잘 보여졌습니다 👍
다만, 카드 메시지 입력이 구현이 안 되어있어서, 이 부분만 수정하면 다음 단계로 넘어가도 되어 보여요.
그 외 코멘트들은 다음 단계 시작하면서 참고해주세요 😊

그리고 useInput 훅 관련해서 답변을 드리자면,
지금 상황에서는 두 훅의 공통된 부분을 하나의 훅으로 만들고 각 사용처에 맞게 확장해 새로운 훅을 만드는 방법도 있을거라 생각해요. 앞으로도 창목님이 고민하신 것처럼 재사용할 수 있지 않을까라는 영역이 계속 생겨날텐데요, 이 부분을 잘 고민해보면서 리팩터링 해나가면 좋아보여요 😊

그리고 로그인 여부에 따른 조건 처리도 잘 이해하신걸로 보입니다 👍

eelkom committed:
feat: OrderPage 내 MessageInput 영역 validate 검사 추가

   - OrderPage의 useInput 훅을 제네릭으로 개선textarea 타입 지원 및 유효성 검사 로직 추가
   - MessageInput, SenderInfo, ReceiverInfo 컴포넌트에서 개선된 useInput 훅 적용
   - validate.ts에 textarea 필드 타입 추가

eelkom commented:
멘토님 자세한 리뷰 감사합니다:)
누락된 카드 메시지 입력 파트 구현 완료 하였고 나머지 부분은 다음 단계에서 반영하겠습니다!

Leejin-Yang commented on src/page/OrderPage/hooks/useInput.ts:
```diff
@@ -0,0 +1,35 @@
+
+type FieldType = 'email' | 'password' | 'text' | 'phoneNumber' | 'textarea';
+
+export type UseInputReturn<T> = {
```
오 제네릭을 활용주셨네요 👍

Leejin-Yang commented on src/page/OrderPage/index.tsx:
```diff
@@ -0,0 +1,63 @@
+  const { id } = useParams<{ id: string }>();
+  const index = Number(id);
+
+  const messageInputText = useInput<HTMLTextAreaElement>('textarea', '축하해요.');
```
기본 메시지 값도 데이터에서 가져올 수 있어요 😊

Leejin-Yang reviewed (approved):
이번 단계 고생 많으셨습니다 👍

이번 단계를 계기로 앞으로도 지금처럼 재사용과 분리에 대해 고민하면 좋아보여요 😊

다음 단계에서 보아요 🙌
